fv3core/stencils/d_sw.py

```diff
@@ -105,64 +190,83 @@ def not_inlineq_pressure(gx: sd, gy: sd, rarea: sd, fx: sd, fy: sd, pt: sd, delp
         pt[0, 0, 0] = pt / delp
 
 
-@gtstencil()
-def ke_from_bwind(ke: sd, ub: sd, vb: sd):
-    with computation(PARALLEL), interval(...):
-        ke[0, 0, 0] = 0.5 * (ke + ub * vb)
+@gtscript.function
+def ub_from_vort(vort, ub):
+    return vort - vort[1, 0, 0]
 
 
-@gtstencil()
-def ub_from_vort(vort: sd, ub: sd):
-    with computation(PARALLEL), interval(...):
-        ub[0, 0, 0] = vort - vort[1, 0, 0]
+@gtscript.function
+def vb_from_vort(vort, vb):
+    return vort - vort[0, 1, 0]
 
 
 @gtstencil()
-def vb_from_vort(vort: sd, vb: sd):
+def ub_vb_zrat_vort(
+    vort: FloatField,
+    ub: FloatField,
+    vb: FloatField,
+    wk: FloatField,
+    f0: FloatField,
+    z_rat: FloatField,
+    kinetic_energy_franction_to_damp: float,
+    dcon_threshold: float,
+    do_f3d: bool,
+    hydrostatic: bool,
+):
+    from __externals__ import local_ie, local_is, local_je, local_js
+
     with computation(PARALLEL), interval(...):
-        vb[0, 0, 0] = vort - vort[0, 1, 0]
+        with horizontal(region[local_is : local_ie + 2, local_js : local_je + 2]):
+            if kinetic_energy_franction_to_damp > dcon_threshold:
+                ub = ub_from_vort(vort, ub)
+                vb = vb_from_vort(vort, vb)
+        if do_f3d and not hydrostatic:
+            vort = zrat_vorticity(wk, f0, z_rat)
+        else:
+            vort = wk[0, 0, 0] + f0[0, 0, 0]
 
 
-@gtstencil()
-def u_from_ke(ke: sd, vt: sd, fy: sd, u: sd):
-    with computation(PARALLEL), interval(...):
-        u[0, 0, 0] = vt + ke - ke[1, 0, 0] + fy
+@gtscript.function
+def u_from_ke(ke, vt, fy):
+    return vt + ke - ke[1, 0, 0] + fy
+
+
+@gtscript.function
+def v_from_ke(ke, ut, fx):
+    return ut + ke - ke[0, 1, 0] - fx
 
 
 @gtstencil()
-def v_from_ke(ke: sd, ut: sd, fx: sd, v: sd):
+def u_and_v_from_ke(
+    ke: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    fx: FloatField,
+    fy: FloatField,
+    u: FloatField,
+    v: FloatField,
+):
+    from __externals__ import local_ie, local_is, local_je, local_js
+
     with computation(PARALLEL), interval(...):
-        v[0, 0, 0] = ut + ke - ke[0, 1, 0] - fx
+        with horizontal(region[local_is : local_ie + 1, local_js : local_je + 2]):
+            u = u_from_ke(ke, vt, fy)
+        with horizontal(region[local_is : local_ie + 2, local_js : local_je + 1]):
```

Comment:
I think these regions would not be necessary if combined with the heat_from_damping stencil, which will only be possible when we can call delnflux from a stencil. If you agree, it would be cool to add a note saying as much, something like #TODO: may be able to remove local regions once heat from_damping is in the same stencil

Reply:
Yep I agree with what you said!